### ISSUE TITLE
Fetch API cannot load file:///xxxxx/assets/FontManifest.json. URL scheme “file” is not supported  (on android device)

### ISSUE THREAD

**Issue by JiaoLiu** [flutter/flutter#128760]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. run flutter build web.
2. copy build/web dir to android phone.
3. use a webview load web/index.html

### Expected results

show the main.dart screen.

### Actual results

inspect the device,  found error "Fetch API cannot load file:///xxx/assets/FontManifest.json. URL scheme “file” is not supported"
& blank screen.

build web under flutter 3.7.0 everything is ok.
flutter 3.10.x error & not work.

### Code sample

<details><summary>Code sample</summary>

```dart
[Paste your code here]
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[Paste your output here]
```

</details>

**Comment by JiaoLiu** [comment]
> 

we are not using webview_flutter. we use native webview to load the html source with 'file://' scheme.  The web source is stored in the sandbox(ios)&SD card(android).

**Comment by huycozy** [comment]
Can you please share a minimal sample code for android native part? (The flutter web build output can be skipped as I can do it)

It would be appreciated if you could provide a GitHub repository in this case. Thank you!

**Comment by JiaoLiu** [comment]
> Can you please share a minimal sample code for android native part? (The flutter web build output can be skipped as I can do it)
> 
> It would be appreciated if you could provide a GitHub repository in this case. Thank you!

 this.webview?.loadUrl(Uri.fromFile(File("file:///sdcard/index.html"))

**Comment by huycozy** [comment]
Unfortunately, I'm hitting `ERR_ACCESS_DENIED` error when doing this. Below is my settings to load web on Android:

```kotlin
        val flutterWebFile = File("${extStoragePath}/web/index.html")
        Log.d("TAG", "flutterWebFile path: ${flutterWebFile.path}") // /storage/emulated/0/web/index.html
        Log.d("TAG", "pathFlutterWeb is existed: ${flutterWebFile.exists()}")   // true
        val pathFlutterWeb = "file:///$flutterWebFile"

        binding.webview.settings.allowFileAccess = true
        binding.webview.settings.allowContentAccess = true
        binding.webview.loadUrl(pathFlutterWeb)
```

Is there missing something to replicate this issue? (checking this on Pixel 3a, Android 12)

**Comment by JiaoLiu** [comment]
> Unfortunately, I'm hitting `ERR_ACCESS_DENIED` error when doing this. Below is my settings to load web on Android:
> 
> ```kotlin
>         val flutterWebFile = File("${extStoragePath}/web/index.html")
>         Log.d("TAG", "flutterWebFile path: ${flutterWebFile.path}") // /storage/emulated/0/web/index.html
>         Log.d("TAG", "pathFlutterWeb is existed: ${flutterWebFile.exists()}")   // true
>         val pathFlutterWeb = "file:///$flutterWebFile"
> 
>         binding.webview.settings.allowFileAccess = true
>         binding.webview.settings.allowContentAccess = true
>         binding.webview.loadUrl(pathFlutterWeb)
> ```
> 
> Is there missing something to replicate this issue? (checking this on Pixel 3a, Android 12)

1、put the builded web source to android device, for example, go to source dir(/xxx/flutter/build/web/), adb push all files to where you will load later:
```
cd /xxx/flutter/build/web/
adb push  *   /storage/emulated/0/Download/
```


2、add`<uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />`to  AndroidManifest.xml

3、 add click method:
```
  if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                when {
                    ContextCompat.checkSelfPermission(
                        MainActivity@this,
                        android.Manifest.permission.READ_EXTERNAL_STORAGE
                    ) == PackageManager.PERMISSION_GRANTED -> {
                        val outdir =    Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS)
                        if (!outdir.exists()) {
                            outdir.mkdirs()
                        }
                        // eg: /storage/emulated/0/Download/
                        this.webview?.loadUrl("file://${File(outdir.absoluteFile,"index.html").absoluteFile}")

                    }
                    shouldShowRequestPermissionRationale(android.Manifest.permission.READ_EXTERNAL_STORAGE) -> {

                }
                    else -> {
                        requestPermissions(
                            arrayOf(android.Manifest.permission.READ_EXTERNAL_STORAGE),
                            0x100)
                    }
                }
            }
```

**Comment by huycozy** [comment]
Thanks but that error still persists with your updated code. I'm checking this on Pixel 3a, Android 12.

<details>
<summary>Screenshot</summary>

![Screenshot 2023-06-16 at 11 36 06](https://github.com/flutter/flutter/assets/104349824/66a0f115-a822-445b-93ff-e78b5bb4ad47)
</details>

I was thinking to storage permission on Android version change, so I checked this on Android 11 device and add `requestLegacyExternalStorage = true` to bypass scope storage if any. But the result is still the same. 

Also, the web path to `index.html` is existing (`path.exists()` is true). I added these settings too, but not working as well:
```kotlin
binding.webview.settings.allowContentAccess = true
binding.webview.settings.allowFileAccess = true
```
Maybe this is a new issue with native webview recently that I'm not aware? 

Can you check this sample code if it's missing anything? [load_flutter_web_128760](https://github.com/huycozy/Android-Native-Sample/tree/load_flutter_web_128760)

**Comment by JiaoLiu** [comment]
> Thanks but that error still persists with your updated code. I'm checking this on Pixel 3a, Android 12.
> 
> Screenshot
> I was thinking to storage permission on Android version change, so I checked this on Android 11 device and add `requestLegacyExternalStorage = true` to bypass scope storage if any. But the result is still the same.
> 
> Also, the web path to `index.html` is existing (`path.exists()` is true). I added these settings too, but not working as well:
> 
> ```kotlin
> binding.webview.settings.allowContentAccess = true
> binding.webview.settings.allowFileAccess = true
> ```
> 
> Maybe this is a new issue with native webview recently that I'm not aware?
> 
> Can you check this sample code if it's missing anything? [load_flutter_web_128760](https://github.com/huycozy/Android-Native-Sample/tree/load_flutter_web_128760)

https://github.com/huycozy/Android-Native-Sample/blob/load_flutter_web_128760/app/build.gradle

line 13.


```
plugins {
    id 'com.android.application'
    id 'org.jetbrains.kotlin.android'
}
android {
    namespace 'com.example.androidnativesample'
    compileSdk 33

    defaultConfig {
        applicationId "com.example.androidnativesample"
        minSdk 24
        targetSdk 33
```

build.gradle  change targetSdk 33 to targetSdk 28.

**Comment by huycozy** [comment]
It's only displaying a white screen instead. In order to make this issue actionable and faster, please provide a sample code and affected Android devices (OS, model) so I can just clone and check it. Thanks!

**Comment by JiaoLiu** [comment]
> It's only displaying a white screen instead. In order to make this issue actionable and faster, please provide a sample code and affected Android devices (OS, model) so I can just clone and check it. Thanks!

can you connect android device to chrome & inspect the web?

maybe you should modify your /web/index.html file.

**Comment by JiaoLiu** [comment]
> It's only displaying a white screen instead. In order to make this issue actionable and faster, please provide a sample code and affected Android devices (OS, model) so I can just clone and check it. Thanks!

can you connect android device to chrome & inspect the web?

maybe you should modify your /web/index.html file.

```

<head>
  <!-- <base href="/"> -->   
</head>

```

**Comment by huycozy** [comment]
When I inspect the webview, I got another error:

```
Access to internal resource at 'file:///storage/emulated/0/Download/manifest.json' from origin 'null' has been blocked by CORS policy: Cross origin requests are only supported for protocol schemes: http, data, chrome, https, chrome-untrusted.
```

<details>
<summary>Inspector</summary>

![Screenshot 2023-06-20 at 11 11 52](https://github.com/flutter/flutter/assets/104349824/c1beef0c-3888-44fa-9d17-f5a5618708ff)
</details>

Can you share a complete and workable sample code along with device info that you are seeing this issue?

**Comment by JiaoLiu** [comment]
> 
you have repoduced this error.  this error will block the first view.  but if you build with flutter 3.7.x, everything is fine.
all android device will show this error, but iphone works.

**Comment by JiaoLiu** [comment]
same flutter code build web.

flutter 3.7.x build web.
![4201687319377_ pic](https://github.com/flutter/flutter/assets/4370982/2bc8cdaa-cb4b-4e40-b94c-65fa97508865)

flutter 3.10.2 build web.
![4211687319377_ pic](https://github.com/flutter/flutter/assets/4370982/9412f22d-6b4c-4f0f-9750-eaba9dec517a)

**Comment by huycozy** [comment]
Unfortunately, I get a similar result as my previous [comment](https://github.com/flutter/flutter/issues/128760#issuecomment-1598089841) when using Flutter 3.7.0 (it's not worked on 3.7.0 as your result). Also, I see a similar error on desktop browser which also uses `file` scheme as mobile.

>Access to XMLHttpRequest at 'file:///Users/huynq/Desktop/file_scheme_128760/build/web/assets/FontManifest.json' from origin 'null' has been blocked by CORS policy: Cross origin requests are only supported for protocol schemes: http, data, isolated-app, chrome-extension, chrome, https, chrome-untrusted.

There is a difference between `XMLHttpRequest` and `internal resource` (my result)/`Fetch API` (your result) in the error message. 

Based on what you share at https://github.com/flutter/flutter/issues/128760#issuecomment-1592301391, it may be related to https://github.com/flutter/engine/pull/39657. Keeping this issue open and labeling it for further investigation.

Sample code: [load_flutter_web_128760](https://github.com/huycozy/Android-Native-Sample/tree/load_flutter_web_128760)

**Comment by yjbanov** [comment]
Is there anything Flutter Web does that prevents this? It seems to be a restriction on Android webivews that disallows loading local files as HTTP requests. Does `fetch` work if you just load a plain HTML+JS page without Flutter Web?

**Comment by JiaoLiu** [comment]
> Unfortunately, I get a similar result as my previous [comment](https://github.com/flutter/flutter/issues/128760#issuecomment-1598089841) when using Flutter 3.7.0 (it's not worked on 3.7.0 as your result). Also, I see a similar error on desktop browser which also uses `file` scheme as mobile.
> 
> > Access to XMLHttpRequest at 'file:///Users/huynq/Desktop/file_scheme_128760/build/web/assets/FontManifest.json' from origin 'null' has been blocked by CORS policy: Cross origin requests are only supported for protocol schemes: http, data, isolated-app, chrome-extension, chrome, https, chrome-untrusted.
> 
> There is a difference between `XMLHttpRequest` and `internal resource` (my result)/`Fetch API` (your result) in the error message.
> 
> Based on what you share at [#128760 (comment)](https://github.com/flutter/flutter/issues/128760#issuecomment-1592301391), it may be related to [flutter/engine#39657](https://github.com/flutter/engine/pull/39657). Keeping this issue open and labeling it for further investigation.
> 
> Sample code: [load_flutter_web_128760](https://github.com/huycozy/Android-Native-Sample/tree/load_flutter_web_128760)


Our Android developer has reviewed your Sample code: [load_flutter_web_128760](https://github.com/huycozy/Android-Native-Sample/tree/load_flutter_web_128760),  and has found that you need to change some WebView settings.

https://github.com/huycozy/Android-Native-Sample/blob/load_flutter_web_128760/app/src/main/java/com/example/androidnativesample/FirstFragment.kt

line 55

```
        binding.btn.setOnClickListener {
            when {
                ContextCompat.checkSelfPermission(
                    requireActivity(),
                    android.Manifest.permission.READ_EXTERNAL_STORAGE
                ) == PackageManager.PERMISSION_GRANTED -> {
                    val outdir =    Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS)
                    if (!outdir.exists()) {
                        outdir.mkdirs()
                    }
                    // eg: /storage/emulated/0/Download/
                    val path = File(outdir.absolutePath,"index.html").absoluteFile
                    Log.d("TAG", "is existed: ${path.exists()}")
                    val webSetting: WebSettings =  binding.webview.getSettings()
                    webSetting.allowContentAccess = true
                    webSetting.javaScriptEnabled = true
                    webSetting.javaScriptCanOpenWindowsAutomatically = true
                    webSetting.allowFileAccess = true
                    webSetting.setSupportZoom(false)
                    webSetting.builtInZoomControls = true
                    webSetting.useWideViewPort = false //##千万不能用
                    webSetting.loadWithOverviewMode = true
                    webSetting.databaseEnabled = true
                    webSetting.domStorageEnabled = true
                    webSetting.setGeolocationEnabled(true)
                    webSetting.pluginState = WebSettings.PluginState.ON
                    webSetting.setRenderPriority(WebSettings.RenderPriority.HIGH)
                    webSetting.cacheMode = WebSettings.LOAD_DEFAULT
                    webSetting.allowFileAccessFromFileURLs = true
                    webSetting.allowUniversalAccessFromFileURLs = true
                    webSetting.savePassword = false
                    webSetting.mediaPlaybackRequiresUserGesture = false
                    WebView.setWebContentsDebuggingEnabled(true)
                    binding.webview.settings.allowContentAccess = true
                    binding.webview.settings.allowFileAccess = true
                    this.binding.webview.loadUrl("file://$path")

                }
                shouldShowRequestPermissionRationale(android.Manifest.permission.READ_EXTERNAL_STORAGE) -> {}
                else -> {
                    requestPermissions(
                        arrayOf(android.Manifest.permission.READ_EXTERNAL_STORAGE),
                        0x100)
                }
            }
        }
```

BTW, you should comment out this line in the index.html file of the website source.

```
<head>
  <!-- <base href="/"> -->   
</head>
```

**Comment by JiaoLiu** [comment]
> Is there anything Flutter Web does that prevents this? It seems to be a restriction on Android webivews that disallows loading local files as HTTP requests. Does `fetch` work if you just load a plain HTML+JS page without Flutter Web?

Fetching to load the file:// schema does not work on the Android WebView, but it works on the iOS WebView.

https://fetch.spec.whatwg.org/

<img width="1353" alt="image" src="https://github.com/flutter/flutter/assets/4370982/d8680dac-7cdf-4799-86a7-7b997c69ad82">

**Comment by yjbanov** [comment]
> https://github.com/huycozy/Android-Native-Sample/blob/load_flutter_web_128760/app/src/main/java/com/example/androidnativesample/FirstFragment.kt

This is not part of the Flutter framework code. I think it's just an example that you can copy and paste into your project and then fix it to make it work. There's nothing the Flutter team can do to fix it.

> Fetching to load the file:// schema does not work on the Android WebView, but it works on the iOS WebView.

I'm still not sure what we'd need to change in Flutter to make it work. Is there a different method for loading the manifest file that Android WebView would accept?

**Comment by JiaoLiu** [comment]
> > https://github.com/huycozy/Android-Native-Sample/blob/load_flutter_web_128760/app/src/main/java/com/example/androidnativesample/FirstFragment.kt
> 
> This is not part of the Flutter framework code. I think it's just an example that you can copy and paste into your project and then fix it to make it work. There's nothing the Flutter team can do to fix it.
> 
> > Fetching to load the file:// schema does not work on the Android WebView, but it works on the iOS WebView.
> 
> I'm still not sure what we'd need to change in Flutter to make it work. Is there a different method for loading the manifest file that Android WebView would accept?

https://github.com/flutter/flutter/issues/128760#issuecomment-1600034131

The same Android code that loads web source built with Flutter 3.7.x will work, but it will not work with Flutter 3.10.x.

I believe the 'main.dart.js' generated by the engine is different.

**Comment by yjbanov** [comment]
One option is to monkey-patch `fetch` to operate on top of `XHR` such that the webview would accept it. I'm going to keep this issue in case we can find a volunteer to implement this.

**Comment by ditman** [comment]
Not sure how current this docs are, but here's some documentation about CORS, and the android web_view:

https://chromium.googlesource.com/chromium/src/+/HEAD/android_webview/docs/cors-and-webview-api.md#file_android_asset_res_urls

**Comment by JiaoLiu** [comment]
> One option is to monkey-patch `fetch` to operate on top of `XHR` such that the webview would accept it. I'm going to keep this issue in case we can find a volunteer to implement this.

Why can't we roll back to using the old way to load the resources(FontManifest.json etc) that were used in Flutter 3.7.0?

**Comment by JiaoLiu** [comment]
> https://chromium.googlesource.com/chromium/src/+/HEAD/android_webview/docs/cors-and-webview-api.md#file_android_asset_res_urls
> 
> (It seems the `file:` needs to be an `asset` or a `resource` of the app for this to work?)


You are correct. We have added `webSetting.allowFileAccessFromFileURLs = true`, so there is no CORS problem.

**Comment by HePeng11** [comment]
guys , any update on this issue ? I also encountered that ~_~.
_"Access to script at 'file:///android_asset/flutter_assets/static/services/xxxxxx.js' from origin 'null' has been blocked by CORS policy: Cross origin requests are only supported for protocol schemes: http, data, chrome, https.", source: file:///android_asset/flutter_assets/static/index.html (0)_

**Comment by JiaoLiu** [comment]
> guys , any update on this issue ? I also encountered that ~_~
> 
> _"Access to script at 'file:///android_asset/flutter_assets/static/services/xxxxxx.js' from origin 'null' has been blocked by CORS policy: Cross origin requests are only supported for protocol schemes: http, data, chrome, https.", source: file:///android_asset/flutter_assets/static/index.html (0)_
> 
> Really appreciate if you comment , it's important for me , thanks
> 
> @JiaoLiu @yjbanov @ditman @huycozy

use flutter branch 3.7.0 to build web
